frontend/src/components/courses/evaluations/Evaluate.tsx

```diff
@@ -0,0 +1,257 @@
+                const comment_data = {
+                    question: id,
+                    content: values.comments[id],
+                }
+                data.comments.push(comment_data);
+            }
+        }
+        for (let id in values.ratings) {
+            if (values.ratings[id] != -1) {
+                const rating_data = {
+                    question: id,
+                    value: values.ratings[id],
+                }
+                data.ratings.push(rating_data);
+            }
+        }
+
+        api.courses
+            .submit(course.id, data)
+            .then(res => {
+                newToast({
+                    message: "A voté !",
+                    level: ToastLevel.Success,
+                });
+
+                setSubmitting(false);
+                setHasVoted(true);
+            })
+            .catch(err => {
+                newToast({
+                    message: err.response.status + " " + err.response.data,
+                    level: ToastLevel.Error,
+                });
+            });
+    }
+
+    const validateComments = (comments, errors) => {
+        for (let i in comments) {
+            let question: Question = getQuestionById(i);
+            let comment: string = comments[i];
+            if (question.required && (comment === "")) {
+                errors[i] = `${question.label} missing required field`;
+                newToast({
+                    message: `La question intitulée : ${question.label} doit être remplit!`,
+                    level: ToastLevel.Error,
+                });
+            }
+        }
+    }
+
+    const validateRatings = (ratings, errors) => {
+        for (let i in ratings) {
+            let question: Question = getQuestionById(i);
+            let rating: number = ratings[i];
+            if (question.required && (rating === -1)) {
+                errors[i] = `${question.label} missing required field`;
+                newToast({
+                    message: `La question intitulée : ${question.label} doit être remplit!`,
+                    level: ToastLevel.Error,
+                });
+            }
+        }
+    }
+
+    const validate = (values) => {
+        {/* Validate the data */ }
+        const errors = {};
+        validateRatings(values.ratings, errors);
+        validateComments(values.comments, errors);
+        return errors;
+    }
+
+    if (hasVoted) return <Redirect to={`/cours/${course.id}`} />
+    return (
+        <Row>
+            <Formik
+                initialValues={initRating(questions)}
+                onSubmit={submitAnswers}
+                validate={validate}
+            >
+                {(props: FormikProps<Values>) => (
+                    <Form onSubmit={props.handleSubmit}>
+                        {questions.map((question: Question) => {
+                            if (question.archived) return null;
+
+                            let field: JSX.Element = <p>Error</p>;
```

Comment:
Don't over-type you variables

Reply:
Use  ternary condition

Reply:
I don't see how the ternary condition would improve the code...